Add support for logging MTimeComparator errors

evhub opened the pull request:
I work for @Yelp. We use `staticconf` to load our configuration files. Sometimes, if something goes wrong, configuration files can disappear. If this happens, we would prefer that error get logged instead of propagating to prevent it from taking down the entire infrastructure. To accomplish this, I have added a function `create_LoggingMTimeComparator(err_logger)` which takes in some logging function and returns an `MTimeComparator` subclass that calls that logging function when a file can't be found instead of propagating an `OSError`.

evhub commented:
Not sure why coveralls is mad at me--the tests are all there.

bchess commented on staticconf/config.py:
```diff
@@ -439,6 +439,29 @@ def has_changed(self):
+        try:
+            return os.path.getmtime(filename)
+        except OSError:
+            err_logger(filename)
```
Maybe also pass the exception just in case that can be used, to check errno, etc

evhub replied:
@bchess `sys.exc_info()` should do that since the function is called within the exception context. And with `sys.exc_info()` you can also see the traceback, not just the exception object. I should probably make it clear that that's the reason the `err_logger` is called within the exception context, though.

bchess replied:
sounds good

bchess commented:
lg2m

dnephin commented on staticconf/config.py:
```diff
@@ -439,6 +439,31 @@ def has_changed(self):
+    """Return an MTimeComparator subclass that calls err_logger(filename)
+    if os.path.getmtime(filename) fails. err_logger is called within the
+    context of the raised OSError, thus sys.exc_info() can be used to
+    examine the exception."""
```
I think there is an easier way to accomplish this.

Modify `MTimeComparator`:

```python

def __init__(self, filenames, compare_func=None):
    self.compare_func = compare_func or os.path.getmtime
```

Then you could implement the rest of this change outside of this repo:

```python

def build_logging_mtime_comparator(err_logger):
    def func(filenames):
        return MTimeComparator(filenames, compare_func=try_getmtime_else_log)
   
    return func

evhub replied:
Using `compare_func` instead of `err_logger` significantly increases the work required to move a service over to the new behavior, which we would like to minimize. I have moved the logging logic inside of MTimeComparator, however (see comment below).

dnephin commented on staticconf/config.py:
```diff
@@ -439,6 +439,31 @@ def has_changed(self):
         return last_mtime < self.last_max_mtime
 
 
+def create_LoggingMTimeComparator(err_logger):
```
This doesn't necessarily need to be part of this repo, but if it stays the pythonic name for this should be something like `build_logging_mtime_comparator()` .  (`build_x` is a commonly used function name in this library)

evhub replied:
Removed `create_LoggingMTimeComparator`.

evhub commented:
I have modified the default behavior of MTimeComparator to catch the error without logging, with the option to provide a logging function. This is a backwards-incompatible change (and would thus require a minor version bump). However, it does not break any existing tests, and we believe it is a more reasonable default than the current behavior of raising an `OSError`. The reasons for this are:

- If a config file disappears inadvertently, we would like to know (hence the logging), but there is no reason it should propagate an error, since it should be able to use the config already loaded into memory.
- If a new version of a program does not need a config file, then we would like to be able to remove that config file at any time so long as we can guarantee that any time the program is run, it will run with the new version. That is not possible if deleting the config file breaks any running old versions.
- If a config file is missing, the error will be identified the next time the program is restarted regardless, so this change does not hide any errors.

If you disagree, I'll change the PR to the minimal behavior you suggested of using a `compare_func`.

dnephin commented:
I'd rather not make a backwards compatible change. I think the current behaviour is a reasonable default. Trying to watch a non-existent config file would be an error.

I understand that logging makes more sense for your use case, but I think it's easy to support that with the change I suggested.

evhub commented:
@dnephin Understood. I modified the PR to only change `MTimeComparator` by adding a `compare_func` option. I would like to keep the logic for creating a `compare_func` from an `err_logger` to reduce the repetition required in our use case, however. However, I have moved that logic to a separate function called `build_compare_func`.

dnephin commented on staticconf/config.py:
```diff
@@ -415,23 +415,44 @@ def has_changed(self):
+            if err_logger is not None:
+                err_logger(filename)
+        return -1
+    return compare_func
```
This function has very little to do with the rest of the module, but I guess it's ok to keep it.

dnephin commented:
Thanks!

evhub commented:
@dnephin Thanks for merging the PR! We would like to make use of this change in our codebase as soon as possible, so we would appreciate it if you could release a new version of PyStaticConfiguration with this change in it. Thanks!
